File: 0225-implement-stack-using-queues/0225-implement-stack-using-queues.py
class MyStack:

    def __init__(self):
        self.a = deque([])
        self.b = deque([])

    def push(self, x: int) -> None:
        self.a.append(x)

    def pop(self) -> int:
        while len(self.a) > 1:
            self.b.append(self.a.popleft())
        self.b, self.a = self.a, self.b
        return self.b.popleft()

    def top(self) -> int:
        while len(self.a) > 1:
            self.b.append(self.a.popleft())
        ret = self.a.popleft()
        self.b.append(ret)
        self.a, self.b = self.b, self.a
        return ret 

    def empty(self) -> bool:
        return len(self.a) + len(self.b) == 0


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: 0225-implement-stack-using-queues/0225-implement-stack-using-queues.java
class MyStack {

    Queue<Integer> q1, q2;
    
    public MyStack() {
        q1 = new LinkedList<>();
        q2 = new LinkedList<>();
    }
    
    public void push(int x) {
        q1.add(x);
    }
    
    public int pop() {
        while (q1.size() != 1) {
            q2.add(q1.poll());
        }
        int ret = q1.poll();
        Queue<Integer> temp = q1;
        q1 = q2;
        q2 = temp;
        return ret;
    }
    
    public int top() {
        while (q1.size() != 1) {
            q2.add(q1.poll());
        }
        int ret = q1.poll();
        q2.add(ret);
        Queue<Integer> temp = q1;
        q1 = q2;
        q2 = temp;
        return ret;
    }
    
    public boolean empty() {
        return q1.size() == 0 && q2.size() == 0;
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack obj = new MyStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * boolean param_4 = obj.empty();
 */